### PR TITLE
+htc #916 add pluggable Transport API to switch out client-side transport layer

### PR DESCRIPTION
**Pull request by jrudolph**
For now, this is basically a refactoring of how the client-side graph is put together.

I had this already lying around for a while half-finished. Thought I could just finish it.

Java APIs are still missing.

Refs #916.

### CODE REVIEW

**Comment by jrudolph** [comment]
Resolved conflicts and compile errors.

**Comment by jrudolph** [comment]
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 1]
```diff
@@ -305,35 +305,45 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
...
+  /**
+   * Similar to `outgoingConnection` but allows to specify a user-defined transport layer to run the connection on.
+   *
+   * Depending on the kind of `ConnectionContext` the implementation will add TLS between the given transport and the HTTP
+   * implementation
+   *
+   * To configure additional settings for requests made using this method,
+   * use the `akka.http.client` config section or pass in a [[akka.http.scaladsl.settings.ClientConnectionSettings]] explicitly.
+   */
+  def outgoingConnectionUsingTransport(
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: good

File: akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import akka.stream.scaladsl.{ Flow, Tcp }
+import akka.util.ByteString
+
+import scala.concurrent.Future
+
+/**
+ * Abstraction to allow the creation of alternative transports to run HTTP on.
+ */
+@ApiMayChange
+trait Transport {
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: Should we mark it akka private for now as well?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: If someone were to implement a transport now they could do it in the akka namespace anyway

File: akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  def connectTo(host: String, port: Int)(implicit system: ActorSystem): Flow[ByteString, ByteString, Future[OutgoingConnection]]
+}
+
+object Transport {
+  def TCP(localAddress: Option[InetSocketAddress], settings: ClientConnectionSettings): Transport =
+    new TCPTransport(localAddress, settings)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: ok

File: akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  private case class TCPTransport(localAddress: Option[InetSocketAddress], settings: ClientConnectionSettings) extends Transport {
+    def connectTo(host: String, port: Int)(implicit system: ActorSystem): Flow[ByteString, ByteString, Future[OutgoingConnection]] =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: `override def` please

File: akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala (hunk 2 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: Hmmm... this is only a **Client** Transport, isn't it?

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: I don't mind, we can move it to `impl` for now and make it `private[akka]`

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: On the other hand, then we cannot expose it in `ConnectionPoolSettings` which is the main purpose of the whole thing.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Transport.scala]: Ah good point on exposing - ok let's keep it public then

**Comment by jrudolph** [comment]
Renamed to ClientTransport and added Java API.

**Pull request merged**